### PR TITLE
Feat validating admission webhook

### PR DESCRIPTION
**Pull request by xuzhu-591**
### Description of your changes

Fixes #

I have:

- [ ] Read and followed Horizon's [contribution process](https://github.com/horizoncd/horizon/CONTRIBUTING.md).
- [ ] Run `make build && make lint` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/config/admission/config.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Resources  []string           `yaml:"resources"`
+	Operations []models.Operation `yaml:"operations"`
+	Versions   []string           `yaml:"versions"`
+}
+
+type Webhook struct {
+	Name           string        `yaml:"name"`
+	Kind           models.Kind   `yaml:"kind"`
+	FailurePolicy  FailurePolicy `yaml:"failurePolicy"`
+	TimeoutSeconds int32         `yaml:"timeoutSeconds"`
```
Comment (iamyeka) [inline pkg/config/admission/config.go:36]: Can we use "time.Duration" as the type of "Timeout" field? Then we can simply use "30s" as the value of the field in the config file, which will be more meaningful.

File: pkg/config/admission/config.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"strings"
+
+	"github.com/horizoncd/horizon/pkg/admission/models"
+)
+
+type FailurePolicy string
+
+func (f FailurePolicy) Eq(other FailurePolicy) bool {
+	return strings.EqualFold(string(f), string(other))
```
Comment (iamyeka) [inline pkg/config/admission/config.go:12]: Why using EqualFold? Any best pratice?

File: pkg/admission/webhook.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+type validateResult struct {
+	err  error
+	resp *Response
+}
+
+type Request struct {
+	Operation    models.Operation       `json:"operation"`
+	Resource     string                 `json:"resource"`
+	ResourceName string                 `json:"resourceName"`
```
Comment (iamyeka) [inline pkg/admission/webhook.go:40]: Confused by the two fields: Resource and ResourceName. Could we choose any other names?

File: pkg/admission/webhook.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	Object       interface{}            `json:"object"`
+	OldObject    interface{}            `json:"oldObject"`
+	Options      map[string]interface{} `json:"options,omitempty"`
+}
+
+type Response struct {
+	Allowed   *bool  `json:"allowed"`
+	Result    string `json:"result,omitempty"`
+	Patch     []byte `json:"patch,omitempty"`
+	PatchType string `json:"patchType,omitempty"`
```
Comment (iamyeka) [inline pkg/admission/webhook.go:52]: What are the optional values for PatchType?

File: pkg/admission/webhook.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+			resCh <- validateResult{nil, response}
+		}(webhook)
+	}
+
+	for res := range resCh {
+		finishedCount++
+		if res.err != nil {
+			return res.err
+		}
+		if res.resp != nil && res.resp.Allowed != nil && !*res.resp.Allowed {
```
Comment (iamyeka) [inline pkg/admission/webhook.go:101]: why use `res.resp.Allowed != nil && !*res.resp.Allowed`?

File: pkg/admission/webhook.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+			log.Infof(ctx, "request denied by webhook: %s", res.resp.Result)
```
Comment (iamyeka) [inline pkg/admission/webhook.go:102]: Is it necessary to print more info like resource, subResource, etc?

File: pkg/admission/webhook.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	herrors "github.com/horizoncd/horizon/core/errors"
+	"github.com/horizoncd/horizon/pkg/admission/models"
+	perror "github.com/horizoncd/horizon/pkg/errors"
+	"github.com/horizoncd/horizon/pkg/util/log"
+)
+
+const DefaultTimeout = 5 * time.Second
```
Comment (iamyeka) [inline pkg/admission/webhook.go:15]: `DefaultTimeout` is not used

File: pkg/admission/webhook.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+func Register(kind models.Kind, webhook Webhook) {
+	switch kind {
+	case models.KindValidating:
+		validatingWebhooks = append(validatingWebhooks, webhook)
+	}
+}
+
+func Clear() {
```
Comment (iamyeka) [inline pkg/admission/webhook.go:28]: `Clear` is not called.

File: pkg/admission/http.go [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+// Get sends the admission request to the webhook server and returns the response
+func (c *HTTPAdmissionClient) Get(ctx context.Context, admitData *Request) (*Response, error) {
+	body, err := json.Marshal(admitData)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", c.config.URL, bytes.NewReader(body))
```
Comment (iamyeka) [inline pkg/admission/http.go:64]: Is it suitable to use `NewRequestWithContext` ?

File: pkg/config/admission/config.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+const (
+	FailurePolicyIgnore FailurePolicy = "ignore"
+	FailurePolicyFail   FailurePolicy = "fail"
+)
+
+type ClientConfig struct {
+	URL      string `yaml:"url"`
+	CABundle string `yaml:"caBundle"`
```
Comment (iamyeka) [inline pkg/config/admission/config.go:22]: No other authentication methods, like bear token,  right ?

File: pkg/config/admission/config.go [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+type Rule struct {
+	Resources  []string           `yaml:"resources"`
+	Operations []models.Operation `yaml:"operations"`
+	Versions   []string           `yaml:"versions"`
+}
+
+type Webhook struct {
+	Name           string        `yaml:"name"`
```
Comment (iamyeka) [inline pkg/config/admission/config.go:33]: Any considerations about using the field `Name`? It look likes it's not printed in any logs. I think that will be a good use case to debug if it's printed in logs.

File: core/middleware/admission/admission.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+			return
+		}
+		attr := record.(auth.AttributesRecord)
+		// non resource request or read only request should be ignored
+		if !attr.IsResourceRequest() || attr.IsReadOnly() {
+			c.Next()
+			return
+		}
+		var object interface{}
+		// read request body and avoid side-effects on c.Request.Body
```
Comment (iamyeka) [inline core/middleware/admission/admission.go:38]: What side-effects?

File: core/middleware/admission/admission.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		if len(bodyBytes) > 0 {
+			contentType := c.ContentType()
+			if contentType == binding.MIMEJSON {
+				if err := json.Unmarshal(bodyBytes, &object); err != nil {
+					response.AbortWithRPCError(c,
+						rpcerror.ParamError.WithErrMsg(fmt.Sprintf("unmarshal request body failed, err: %v", err)))
+					return
+				}
+			} else {
+				log.Warningf(c, "unsupported content type: %s", contentType)
```
Comment (iamyeka) [inline core/middleware/admission/admission.go:55]: Does it need to return here because `object` has no value?

File: pkg/admission/http.go (hunk 9 above)
Comment (iamyeka) [reply pkg/admission/http.go:64]: You use `ctx, cancelFunc := context.WithCancel(ctx)` in the upper method named `Validating`.

File: pkg/admission/webhook.go (hunk 6 above)
Comment (xuzhu-591) [reply pkg/admission/webhook.go:102]: fixed

Comment (xuzhu-591) [reply pkg/admission/webhook.go:102]: added

File: pkg/config/admission/config.go (hunk 1 above)
Comment (xuzhu-591) [reply pkg/config/admission/config.go:36]: fixed

File: pkg/admission/webhook.go (hunk 7 above)
Comment (xuzhu-591) [reply pkg/admission/webhook.go:15]: fixed

File: pkg/admission/webhook.go (hunk 8 above)
Comment (xuzhu-591) [reply pkg/admission/webhook.go:28]: fixed

File: pkg/admission/http.go (hunk 9 above)
Comment (xuzhu-591) [reply pkg/admission/http.go:64]: fixed

File: pkg/admission/webhook.go (hunk 3 above)
Comment (xuzhu-591) [reply pkg/admission/webhook.go:40]: Refer to RequestInfo struct, rename ResourceName to Name.

File: pkg/admission/webhook.go (hunk 4 above)
Comment (xuzhu-591) [reply pkg/admission/webhook.go:52]: removed

File: pkg/config/admission/config.go (hunk 10 above)
Comment (xuzhu-591) [reply pkg/config/admission/config.go:22]: yeah

File: pkg/config/admission/config.go (hunk 11 above)
Comment (xuzhu-591) [reply pkg/config/admission/config.go:33]: removed

File: core/middleware/admission/admission.go (hunk 13 above)
Comment (xuzhu-591) [reply core/middleware/admission/admission.go:55]: return if content type is nusupported.

File: core/cmd/cmd.go [hunk 14]
```diff
@@ -624,7 +631,8 @@ func Init(ctx context.Context, flags *Flags, coreConfig *config.Config) {
 			middleware.MethodAndPathSkipper(http.MethodPost, regexp.MustCompile("^/apis/core/v[12]/users/login"))),
 		prehandlemiddle.Middleware(r, manager),
 		auth.Middleware(rbacAuthorizer, authzSkippers...),
-		tagmiddle.Middleware(), // tag middleware, parse and attach tagSelector to context
+		tagmiddle.Middleware(),
```
Comment (kilosonc) [inline core/cmd/cmd.go:634]: comment?

**Review by kilosonc — APPROVED** [review]

**Pull request merged**